### DISCUSSION TITLE
Library imported server side, but I see client side error

### DISCUSSION THREAD

**Discussion by PaulKushch** [vercel/next.js#19377] [Help] [answered]
# Bug report

## Describe the bug

I have a strange error. When I import mjml-react module (that only works in node.js) in a api route everything works with _next dev_ command, but when I run _next build_ and then _next start_ I get error **Error: Cannot find module '../lib/utils.js'**. When I google around I see that the problem seems to be that mjml-react is not run in node (i.e. it is not possible to run mjml-react client side), but should not a next.js api route be server side in node?

## To Reproduce

I have the code below in _pages/api/customer/customerrequest.js_ file:

```
import generate from '@utilities/customerRequestMail';
import mail from '@utilities/awsMailer';
import { render } from 'mjml-react';

const { db } = require('@/utilities/db');

export default async function handler(req, res) {
  try {
    const { html } = render(generate(), { validationLevel: 'soft' });
    await mail({
      toAdress: 'email@email.com',
      subject: 'Subject',
      text: 'Hello',
      html: html,
    });
    return res.status(200).end();
  } catch (error) {
    return res.status(400).end();
  }
}
```

And the mjml-react code in _pages/api/utilities/customerRequestMail.js_ file:

```
import {
  Mjml,
  ...
} from 'mjml-react';

const generate = () => {
  return (
    <Mjml>
       ...
    </Mjml>
  );
};

export default generate;

```

## Expected behavior

I expect that package imported in api route should be run server side in node.
## System information

- OS: Ubuntu
- Browser (if applies) Chrome
- Version of Next.js: 10.0.0
- Version of Node.js: 12.18.0
- Deployment: localhost. Commands used: next build and then next start.
form]

**Comment by balazsorban44** [comment]
`pages/api/utilities/customerRequestMail.js` is inside your `/api` folder, so it should also be in the form of a Next api route handler. Instead, you are returning JSX from it, like it was a React Component. Try moving this file outside of `api`.


In fact, if I understand your code correctly, you probably want to `/api/utilities` outside all together, if none of those files have the same format as `NextApiHandler`:

```js
export default function(req, res) {
  //...
}
```


Docs: https://nextjs.org/docs/api-routes/introduction

**Comment by PaulKushch** [comment]
@balazsorban44 Thanks for your proposal. If I move _pages/api/utilities/customerRequestMail.js_ outside _/api_ then would it not be bundled as client side code? Any way, I tried but I get the same error _Error: Cannot find module '../lib/utils.js'_.

The general problem seems to be that _MJML uses html-minifier, which uses uglify-js, and uglify-js is not compatible with Webpack._, see here https://github.com/mjmlio/mjml/issues/1365. 

**What I tried now:**
Now I actually just did put the the mjml-react file in same folder as the api route, but the error persists:

_/pages/api/customer/customerRequestMail.js:_
```
import {
  Mjml,
  ...
} from 'mjml-react';

export const generate = () => {
  return (
    <Mjml>
       ...
    </Mjml>
  );
};
```
And my route _/pages/api/customer/customerRequest.js:_

```
import * as custReq from './customerRequestMail';
import mail from '@utilities/awsmailer';
import { render } from 'mjml-react';
const { db } = require('@/utilities/db');

export default async function handler(req, res) {
  try {
    const { html } = render(custReq.generate(), { validationLevel: 'soft' });
    await mail({
      toAdress: 'mail@email.com',
      subject: 'Subject',
      text: 'Hello',
      html: html,
    });
    return res.status(200).end();
  } catch (error) {
    return res.status(400).end();
  }
};
```

This is the exact same pattern as _mjml-react_ shows in their example repo at _https://github.com/wix-incubator/mjml-react-example/tree/master/src_
I find it very strange that I get this eror **only** after I run **next build** and then **next start**, because it works with **next dev** command.
Best Regards

**Comment by balazsorban44** [comment]
No, moving your code outside `/api`, in itself, has nothing to do with if it will be included in the client bundle or not.

The example you link to is using `express`, not `next`.

You can think of the `pages/api` folder as routes inside an `express` app. Each file in `api` represents a route.

Code that is explicitely imported/used inside files inside `pages/api`, `getStaticProps` `getStaticPaths` `getServerSideProps` will be excluded from client-side bundles.

An example:

```js
// pages/index.js
import foo from "bar"


export default function Home() {
  return null
}


export async function getStaticProps() {
  foo() // the code related to this will only be included in the server bundle.
  return {
    props: {}
  }
}
```

Here, the module "bar" will only be included server-side.


I might have not been a good enough answer. Be sure that your utilities folder is outside of `pages` as well! Everything inside pages will become a page, everything inside pages/api will become an api route (using a single default exported function as api handler)

Something like this should work.
![image](https://user-images.githubusercontent.com/18369201/99877727-80f39000-2c00-11eb-9d8f-7734bb16fe85.png)

Where you only import `customerRequestMail.js` in`pages/api/customer/customerRequest.js`

**Comment by PaulKushch** [comment]
@balazsorban44 Thanks, I maybe was unclear. I tried all your suggestions but the problem still persists (I now did move utilities to top level of working directory). Issue seems to be that MJML uses html-minifier, which uses uglify-js, and uglify-js is not compatible with Webpack., see here mjmlio/mjml#1365.

So it will not help moving files outside _/api_. But I wounder why it works in dev but not after build. Any ideas how to solve that?

Best regards

**Comment by balazsorban44** [comment]
A clear reproduction repository and a link to it would be helpful to help you debug.

**Comment by PaulKushch** [comment, marked as answer]
@balazsorban44 I resolved the issue :)

next.config.js must be adjusted:

```
const path = require('path');
const withBundleAnalyzer = require('@next/bundle-analyzer')({
  enabled: process.env.ANALYZE === 'true',
});
module.exports = withBundleAnalyzer({
  webpack: (config, { buildId, dev, isServer, defaultLoaders, webpack }) => {
    // Note: we provide webpack above so you should not `require` it
    // Perform customizations to webpack config
    config.plugins.push(new webpack.IgnorePlugin(/^pg-native$/));
    config.module.rules.push({
      test: path.resolve(__dirname, 'node_modules/uglify-js/tools/node.js'),
      loader: 'null-loader',
    });

    // Important: return the modified config
    return config;
  },
  target: 'serverless', //
  ...
```
So _null-loader_ must be used. 
So this has nothing to do with where the files are placed :)

Best Regards

**Comment by balazsorban44** [comment]
So this can be closed, I guess?

**Comment by PaulKushch** [comment]
@balazsorban44 Yes, but please give me a day to test before close. Thanks for you help mate :+1:

**Comment by balazsorban44** [comment]
I am not a maintainer nor the creator of the issue. I cannot close it, but you can, if you think you found the solution.

Happy that I could help.
